core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -109,20 +116,21 @@ private[spark] class OutputCommitCoordinator(conf: SparkConf, isDriver: Boolean)
    * @param maxPartitionId the maximum partition id that could appear in this stage's tasks (i.e.
    *                       the maximum possible value of `context.partitionId`).
    */
-  private[scheduler] def stageStart(stage: StageId, maxPartitionId: Int): Unit = synchronized {
+  private[scheduler] def stageStart(stage: Int, maxPartitionId: Int): Unit = synchronized {
     stageStates(stage) = new StageState(maxPartitionId + 1)
```

Comment:
My memory is a bit rusty here, but are we changing the semantics of which task can commit here ?
Couple of queries:
* Are we allowing task from a previous stage attempt to commit for current stage attempt ?
** If yes, we should not overwrite `stageStates(stage)` if it exists.
*** Based on `TaskIdentifier` above, I this yes ?
** If no, we should check and reject commit requests from tasks from 'older' stage when the current stage attempt is different.

Reply:
I don't think the semantics are changing. It's always been racy, in that either of the concurrent tasks from different stage attempts may succeed first. And I'm almost sure the assumption is that both task attempts are equivalent (i.e. the output is deterministic or at least should be), so it should be fine for either to be committed.

The problem is that without this change the coordinator would allow both attempts to commit, and that is kinda bad.

Reply:
There are two cases here (both not handled in existing/earlier code).

Handled in PR:
* Stage S1 attempt A1 launched.
* Tasks T1_1 launched for partition P1
* A1 fails
* Stage S1 attempt A2 launched.
* Tasks T1_2 for partition P1 launched.
* T1_1 finishes, and is allowed to commit.

IMO not handled in PR:
* Stage S1 attempt A1 launched.
* Tasks T1_1.1 launched for partition P1
* Tasks T1_1.2 launched for partition P1 (speculative)
* Task T1_1.1 committed.
* A1 fails
* Stage S1 attempt A2 launched for some other pending partitions.
* Tasks T1_1.2 wants to commit.

T1_1.2 will be allowed to commit.
Now we have two tasks for same partition successfully committing.

Did I miss something here ?

Reply:
If I read @vanzin's PR right, T1_1.2 will be allowed to commit - since there is a stageEnd + stageStart in between (which clear the existing stage state).

Reply:
I checked the related code in `DAGScheduler`, if `T1_1.1` succeeds, the re-tried stage won't launch task for this partition, because Spark tracks finished tasks for a job.

Reply:
Ah, I see your confusion.
Please note that T1_1.2 is a speculative task, not a retried task.

Reply:
Actually @mridulm have a point here, on Scenario 2 we'll finally have  two attempts of a single task committed. Normally this shall not cause any problem because the two attempts shall generate the same output. But I'm still wondering whether it will generate corrupted data when the commit process of T1_1.1 didn't finish (so it may write the same dir with T1_1.2 at the same time) or T1_1.2 didn't finish the commit process(the executor get killed in the mid of commit process).

Reply:
Yeah, I think this can happen. The problem is that with the current way it's used, the output committer forgets the commit status between stage retries. I think the right thing would be for the committer to keep the stage-related stage until the scheduler is done with all its attempts.

> whether it will generate corrupted data when the commit process of T1_1.1 didn't finish

I don't think that's the problem. The problem is that if both the initial task and the speculative task finish successfully, but across stage attempt barriers (so the output committer is "reset" in between), both will be allowed to commit, so you get duplicate data.

Reply:
So in scenario 2, once the first task finishes and is committed, the taskset manager will kill the speculative task T1_1.2. But since it sends an async message to kill the task, the task could actually try to commit after another task fails and causes the stage to remove itself from the output commit coordinator and after it starts another stage attempt.  So it could actually end up committed the task output for T1_1.2.  I'm not sure this case by itself is a problem though since if it actually committed T1_1.1 and T1_1.2 is allowed to commit, they should have the same output and commitJob would handle in at least most cases.   The caveat there though would be if since T1_1.1 was committed, the second stage attempt could finish and call commitJob while T1_1.2 is committing since spark thinks it doesn't need to wait for T1_1.2.  Anyway this seems very unlikely but we should protect against it.

There is another case though here where T1_1.1 could have just asked to be committed, but not yet committed, then if it gets delayed committing, the new stage attempt starts and T1_1.2 asks if it could commit and is granted, so then both try to commit at the same time causing corruption.

Reply:
>  I think the right thing would be for the committer to keep the stage-related state until the scheduler is done with all its attempts.

We should change the `DAGScheduler` a little bit that, if a stage is killed and going to be re-tried, do not clear the stage states in output coordinator.

Reply:
Yep, I'm going down that path. I just want to add a proper test to make sure the behavior is correct and that's a little bit tricky.

Reply:
I agree @vanzin, @cloud-fan. We should remove the stage info only after the stage is done.